u_parrots.go

```diff
@@ -619,38 +633,86 @@ func (uconn *UConn) generateRandomizedSpec(WithALPN bool) (ClientHelloSpec, erro
 		}
 		p.Extensions = append(p.Extensions, &ks, &pskExchangeModes, &supportedVersionsExt)
 	}
-	err = shuffleTLSExtensions(p.Extensions)
-	if err != nil {
-		return p, err
-	}
+	r.shuffleTLSExtensions(p.Extensions)
 	err = uconn.SetTLSVers(p.TLSVersMin, p.TLSVersMax)
 	if err != nil {
 		return p, err
 	}
 
-	return p, nil
+	return p, r.savedErr
+}
+
+// utlsRandomizer never returns errors, but stores first one in savedErr
+type utlsRandomizer struct {
+	r        io.Reader
+	savedErr error
+}
+
+func makeUtlsRandomizer(seed uint64) *utlsRandomizer {
+	seedBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(seedBytes, seed)
+	tdHkdf := hkdf.New(sha256.New, seedBytes, []byte("uTLS salty salt"), nil)
+	return &utlsRandomizer{r: tdHkdf}
+}
+
+// randUInt64 is the only function that touches source of entropy,
+// and other primitives are built using randUInt64
+func (r *utlsRandomizer) randUInt64() uint64 {
+	uintBytes := make([]byte, 8)
+	_, err := r.r.Read(uintBytes)
+	if r.savedErr == nil && err != nil {
+		r.savedErr = err
+	}
+	return binary.BigEndian.Uint64(uintBytes)
+}
+
+// returns int64 in [0, max)
+func (r *utlsRandomizer) randInt64n(max int) int64 {
+	if max == 0 {
+		return 0
+	}
+	randUint := r.randUInt64() % uint64(max)
+	if randUint > math.MaxInt64 {
+		randUint >>= 1
+	}
+	return int64(randUint)
+}
+
+// returns int in [0, max) up to max int32
+func (r *utlsRandomizer) randIntn(max int) int {
```

Comment:
Use the runtime libray's implementation.

Output is non-uniform for most values of `n`.